src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs

```diff
@@ -779,5 +779,43 @@ public async Task DynamicFilesNotPassedToSourceGenerators()
             Assert.True(noTreesPassed.HasValue);
             Assert.False(noTreesPassed!.Value);
         }
+
+        [Fact]
+        public async Task LinkedDocumentOfFrozenShouldNotRunSourceGenerator()
+        {
+            using var workspace = CreateWorkspaceWithPartialSemantics();
+            var generatorRan = false;
+            var analyzerReference = new TestGeneratorReference(new CallbackGenerator(_ => { }, onExecute: _ => { generatorRan = true; }, source: "// Hello World!"));
+
+            var originalDocument1 = AddEmptyProject(workspace.CurrentSolution, name: "Project1")
+                .AddAnalyzerReference(analyzerReference)
+                .AddDocument("RegularDocument.cs", "// Source File", filePath: "RegularDocument.cs");
+
+            // this is a linked document of document1 above
+            var originalDocument2 = AddEmptyProject(originalDocument1.Project.Solution, name: "Project2")
+                .AddAnalyzerReference(analyzerReference)
+                .AddDocument(originalDocument1.Name, await originalDocument1.GetTextAsync().ConfigureAwait(false), filePath: originalDocument1.FilePath);
+
+            var project2 = originalDocument2.Project;
+            Assert.True(workspace.SetCurrentSolution(_ => project2.Solution, WorkspaceChangeKind.SolutionChanged));
+
+            // check both documents are in the snapshot
+            Assert.Equal(project2.GetRequiredDocument(originalDocument2.Id).GetLinkedDocumentIds().Single(), originalDocument1.Id);
+            Assert.False(generatorRan);
+
+            // froze document2 
+            var documentIdsToTest = new[] { originalDocument1.Id, originalDocument2.Id };
+            foreach (var documentIdToTest in documentIdsToTest)
+            {
+                var frozenDocument = workspace.CurrentSolution.GetRequiredDocument(documentIdToTest).WithFrozenPartialSemantics(CancellationToken.None);
```

Comment:
Oh this isn't actually testing what we wanted anymore -- we should only call WithFrozenPartialSemantics once outside the loop -- we want to assert that the document we didn't call freeze on is still frozen too.